internal/plugin/connectors/tcp/mssql/connector_test.go

```diff
@@ -1,59 +1,372 @@
+				opt.ServerLoginResponse = expectedLoginResponse
+			},
+		),
 	)
-	actualBackendConn, err := connector.Connect(clientConn, creds)
 
+	_, _ = runDefaultTestConnect(connector)
+	expectedClient := connector.clientConn
+
+	assert.Equal(t, actualLoginResponse, expectedLoginResponse)
+	// confirm that WriteLoginResponse is called with the client connection
+	assert.Equal(t, expectedClient, actualClient)
+}
+
+func TestWriteLoginResponseFails(t *testing.T) {
+	methodFails(t, func(connectorOptions *types.ConnectorOptions) {
+		connectorOptions.WriteLoginResponse = func(
+			w io.ReadWriteCloser,
+			loginRes *mssql.LoginResponse,
+		) error {
+			return methodFailsExpectedErr
+		}
+	})
+}
+
+func TestProductionWriteErr(t *testing.T) {
+	// production version of WriteError
+	var writeError types.WriteErrorFunc = mssql.WriteError72
+
+	// expected error
+	expectedErr := mssql.Error{
+		Number:     1,
+		State:      2,
+		Class:      3,
+		Message:    "test-message",
+		ServerName: "test-server-name",
+		ProcName:   "test-proc-name",
+		LineNo:     4,
+	}
+
+	r, w := net.Pipe()
+	go func() {
+		writeError(w, expectedErr)
+	}()
+
+	// production read error
+	actualErr, err := mssql.ReadError(r)
 	assert.Nil(t, err)
 	if err != nil {
 		return
 	}
 
-	assert.Equal(t, expectedBackendConn, actualBackendConn)
+	assert.Contains(t, actualErr.Error(), expectedErr.Error())
+}
+
+// Test helpers
+//
+
+// runDefaultTestConnect passes in default values to the Connect method of a connector.
+// This helps avoid boilerplate
+func runDefaultTestConnect(
+	connector *SingleUseConnector,
+) (net.Conn, error) {
+	clientConn := mock.NewNetConn(nil)
+	creds := pluginconnector.CredentialValuesByID{
+		"credName": []byte("secret"),
+	}
+
+	return connector.Connect(clientConn, creds)
+}
+
+// methodFails checks that the expected error is present in:
+// 1. the error returned from the call to the Connect method
+// 2. the error propagated to the client
+var methodFailsExpectedErr = errors.New("failed for the test")
+
+func methodFails(
+	t *testing.T,
+	connectorOption types.ConnectorOption,
+) {
+	var actualClientErr error
+	var actualClient io.ReadWriteCloser
+	var actualErr error
+
+	// expected error on method
+	expectedErr := methodFailsExpectedErr
+
+	connector := NewSingleUseConnectorWithOptions(
+		mock.DefaultConnectorOptions(),
+		func(connectorOptions *types.ConnectorOptions) {
+			// error should always be written to client
+			connectorOptions.WriteError = func(w io.ReadWriteCloser, err mssql.Error) error {
+				actualClient = w
+				actualClientErr = err
+				return nil
+			}
+		},
+		connectorOption,
+	)
+
+	_, actualErr = runDefaultTestConnect(connector)
+	expectedClient := connector.clientConn
+
+	// confirms error returned by #Connect contains expected error
+	assert.Equal(t, expectedErr, errorspkg.Cause(actualErr))
+	// confirms that errors are always written to the client and not something else
+	assert.Equal(t, expectedClient, actualClient)
```

Comment:
Is it possible for the errors to be written to something else besides the client?